src/client/context.cc

```diff
@@ -385,6 +371,50 @@ void ClientContext::UpdateClientTunnel(
   }
 }
 
+bool ClientContext::AddServerTunnel(
+    const TunnelAttributes& tunnel,
+    bool is_http)
+{
+  auto local_destination = LoadLocalDestination(tunnel.keys, true);
+  auto server_tunnel =
+      is_http ? std::make_unique<kovri::client::I2PServerTunnelHTTP>(
+                    tunnel, local_destination)
+              : std::make_unique<kovri::client::I2PServerTunnel>(
+                    tunnel, local_destination);
+  // Insert server tunnel
+  if (!InsertServerTunnel(
+          local_destination->GetIdentHash(), std::move(server_tunnel)))
+    {
+      LOG(error) << "Instance: server tunnel for destination "
+                 << GetAddressBook().GetB32AddressFromIdentHash(
+                        local_destination->GetIdentHash())
+                 << " already exists";
+      return false;
+    }
+  // Start the new server tunnel
+  std::lock_guard<std::mutex> lock(m_ServerMutex);
+  m_ServerTunnels[local_destination->GetIdentHash()]->Start();
+  return true;
+}
+
+bool ClientContext::AddClientTunnel(const TunnelAttributes& tunnel)
+{
+  std::shared_ptr<kovri::client::ClientDestination> local_destination;
+  if (!tunnel.keys.empty())
+    local_destination = LoadLocalDestination(tunnel.keys, false);
+  // Insert client tunnel
+  auto client_tunnel = std::make_unique<kovri::client::I2PClientTunnel>(
+      tunnel, local_destination);
+  if (!InsertClientTunnel(tunnel.port, std::move(client_tunnel)))
+    {
+      LOG(error) << "Instance: failed to insert new client tunnel";
+      return false;
+    }
+  std::lock_guard<std::mutex> lock(m_ClientMutex);
+  m_ClientTunnels[tunnel.port]->Start();
```

Comment:
This is the only logic change. ClientContext::AddClientTunnel is called from 2 places : one of them didn't have the Start call.  I don't know if it should be called, but i found it weird that they didn't do the same.